Lua: allow to pass custom reader options to `pandoc.read`

tarleb opened the pull request:
Reader options can now be passed as an optional third argument to
`pandoc.read`. The object can either be a table or
`PANDOC_READER_OPTIONS`; the latter ensures that the same options used
for main input parsing are re-used to parse the input to `pandoc.read`.

Closes: #7656

tarleb commented:
Might need a little more work, e.g. documentation. I'm also not sure whether we should provide a `pandoc.ReaderOptions` constructor and require the argument to have that type:

```lua
local opts = pandoc.ReaderOptions {tab_stops = 2}    -- or, e.g., pandoc.ReaderOptions(PANDOC_READER_OPTIONS)
return pandoc.read('*CommonMark*', 'commonmark', opts)
```

One advantage of that would be that it is easier to document.

jgm commented:
I suppose another advantage of pandoc.ReaderOptions would be that you could throw an error on unknown options (giving them feedback if they get the field names wrong).

The important thing is to get as much consistency as possible in the interface.  E.g., if you have to do something similar with Attributes, and can't just pass in a table, it could be a reason to do the same here.

tarleb commented:
Most things are very permissive: whenever an `Attr`, `Attributes`, or `ListAttributes` is expected, it's also possible to pass in an equivalent table or even just an ID string. A Lua string can be used as a plain string, as `Str`, or as `Plain [Str _]`. If a list is expected, we generally allow a single item instead, which is then treated as a singleton list. So I'd lean towards at least allowing tables to be used instead of full `pandoc.ReaderOptions` elements.

On the other hand, we don't support that yet for Citations, but using `{id='einstein1905', mode='NormalCitation'}` could be convenient at times. Maybe I should open a new issue for this question.

jgm commented:
That's okay with me -- should I merge this then?

tarleb commented:
I've added a `pandoc.ReaderOptions` constructor and changed the code towards a better and more maintainable method of marshalling Lua tables into ReaderOptions tables.

Docs need updating, but good to merge otherwise.

jgm commented:
I'll keep this open pending documentation (otherwise we're more likely to forget).

tarleb commented:
OK, done.

jgm commented:
Question:  Is there a reason there's not a Pushable instance for ReaderOptions?  If there were, I could just have ReaderOptions be the optional second parameter of Reader (in custom readers) rather than using the gloabl PANDOC_READER_OPTIONS.

tarleb commented:
> Question:  Is there a reason there's not a Pushable instance for ReaderOptions?  If there were, I could just have ReaderOptions be the optional second parameter of Reader (in custom readers) rather than using the gloabl PANDOC_READER_OPTIONS.

Not really.  I wanted to avoid any confusion on whether `push` would act
like `pushReaderOptions` or like `pushReaderOptionsReadonly`, but the
expectation would probably to have a modifiable object.

Having a second parameter seems like the nicer interface. I'll add a
Pushable instance in a minute.
